File: test-suite/handwritten-src/swift/ProviderTest.swift
import DjinniSupport
import XCTest
@testable import TestSuite

final class ProviderTest: XCTestCase {
    func test() throws {
        let providerStr = try TestProvider_statics.getProviderString()
        XCTAssertEqual(providerStr(), "hello")
        // C++ provider can be called more than once
        XCTAssertEqual(providerStr(), "hello")
        XCTAssertEqual(providerStr(), "hello")

        let providerInt = try TestProvider_statics.getProviderInt()
        XCTAssertEqual(providerInt(), 42)
        XCTAssertEqual(providerInt(), 42)

        let swiftProviderStr = Provider { "world" }
        XCTAssertEqual(try TestProvider_statics.callProviderString(swiftProviderStr), "world")
        // Host provider can be used in multiple C++ calls
        XCTAssertEqual(try TestProvider_statics.callProviderString(swiftProviderStr), "world")

        let swiftProviderInt = Provider { 123 as Int32 }
        XCTAssertEqual(try TestProvider_statics.callProviderInt(swiftProviderInt), 123)
        XCTAssertEqual(try TestProvider_statics.callProviderInt(swiftProviderInt), 123)

        let np = try TestProvider_statics.getNestedProvider()
        XCTAssertEqual(np.s(), 99)
        XCTAssertEqual(try TestProvider_statics.callNestedProvider(NestedProvider(s: Provider { 55 })), 55)

        let providerObj = try TestProvider_statics.getProviderObject()
        let obj = providerObj()
        XCTAssertEqual(try obj.getValue(), 42)
        XCTAssertEqual(try obj.getName(), "expensive")
        XCTAssertEqual(try TestProvider_statics.callProviderObject(providerObj), 42)
        XCTAssertEqual(try TestProvider_statics.callProviderObjectGetName(providerObj), "expensive")

        var creationCount = 0
        let swiftProviderObj = Provider<SimpleObject> {
            creationCount += 1
            return try! TestProvider_statics.createSimpleObject(999, name: "swift-created")
        }
        XCTAssertEqual(creationCount, 0)
        // C++ calls host provider twice to verify multi-call; expect 2 creations
        XCTAssertEqual(try TestProvider_statics.callProviderObject(swiftProviderObj), 999)
        XCTAssertEqual(creationCount, 2)
        XCTAssertEqual(try TestProvider_statics.callProviderObjectGetName(swiftProviderObj), "swift-created")
        XCTAssertEqual(creationCount, 4)

        let npi = try TestProvider_statics.getNestedProviderInterface()
        XCTAssertEqual(try npi.obj().getValue(), 777)
        XCTAssertEqual(try TestProvider_statics.callNestedProviderInterface(
            NestedProviderInterface(obj: Provider {
                try! TestProvider_statics.createSimpleObject(888, name: "nested-swift")
            })
        ), 888)
    }
}

File: test-suite/handwritten-src/ts/ProviderTest.ts
import { TestCase, allTests, assertEq } from "./testutils";
import * as test from "../../generated-src/ts/test";

class ProviderTest extends TestCase {
  m: test.Test_statics;
  constructor(module: test.Test_statics) {
    super(module);
    this.m = module;
  }

  test() {
    var providerStr = this.m.testsuite.TestProvider.getProviderString();
    var result = providerStr();
    assertEq(result, "hello");
    // C++ provider can be called more than once
    assertEq(providerStr(), "hello");
    assertEq(providerStr(), "hello");

    var providerInt = this.m.testsuite.TestProvider.getProviderInt();
    assertEq(providerInt(), 42);
    assertEq(providerInt(), 42);

    var jsProviderStr = () => "world";
    assertEq(this.m.testsuite.TestProvider.callProviderString(jsProviderStr), "world");
    // Host provider can be used in multiple C++ calls
    assertEq(this.m.testsuite.TestProvider.callProviderString(jsProviderStr), "world");

    var jsProviderInt = () => 123;
    assertEq(this.m.testsuite.TestProvider.callProviderInt(jsProviderInt), 123);
    assertEq(this.m.testsuite.TestProvider.callProviderInt(jsProviderInt), 123);

    var np = this.m.testsuite.TestProvider.getNestedProvider();
    assertEq(np.s(), 99);

    assertEq(this.m.testsuite.TestProvider.callNestedProvider({ s: () => 55 }), 55);

    var providerObj = this.m.testsuite.TestProvider.getProviderObject();
    var obj = providerObj();
    assertEq(obj.getValue(), 42);
    assertEq(obj.getName(), "expensive");

    var result1 = this.m.testsuite.TestProvider.callProviderObject(providerObj);
    assertEq(result1, 42);

    var result2 = this.m.testsuite.TestProvider.callProviderObjectGetName(providerObj);
    assertEq(result2, "expensive");

    var creationCount = 0;
    var jsProviderObj = () => {
      creationCount++;
      return this.m.testsuite.TestProvider.createSimpleObject(999, "js-created");
    };

    assertEq(creationCount, 0);

    // C++ calls host provider twice to verify multi-call; expect 2 creations
    var result3 = this.m.testsuite.TestProvider.callProviderObject(jsProviderObj);
    assertEq(result3, 999);
    assertEq(creationCount, 2);

    var result4 = this.m.testsuite.TestProvider.callProviderObjectGetName(jsProviderObj);
    assertEq(result4, "js-created");
    assertEq(creationCount, 4);

    var npi = this.m.testsuite.TestProvider.getNestedProviderInterface();
    var npiObj = npi.obj();
    var npiValue = npiObj.getValue();
    assertEq(npiValue, 777);

    var result5 = this.m.testsuite.TestProvider.callNestedProviderInterface({
      obj: () => this.m.testsuite.TestProvider.createSimpleObject(888, "nested-js"),
    });
    assertEq(result5, 888);

    var memoCreationCount = 0;
    var unmemoizedProvider = () => {
      memoCreationCount++;
      return this.m.testsuite.TestProvider.createSimpleObject(555, "memoized");
    };

    function memoize<T>(fn: () => T): () => T {
      let cached: { value: T } | null = null;
      return () => {
        if (cached === null) {
          cached = { value: fn() };
        }
        return cached.value;
      };
    }

    var memoizedProvider = memoize(unmemoizedProvider);

    assertEq(memoCreationCount, 0);
    var result6 = this.m.testsuite.TestProvider.callProviderObject(memoizedProvider);
    assertEq(result6, 555);
    assertEq(memoCreationCount, 1);

    var result7 = this.m.testsuite.TestProvider.callProviderObjectGetName(memoizedProvider);
    assertEq(result7, "memoized");
    assertEq(memoCreationCount, 1);
  }
}

allTests.push(ProviderTest);
